src/ucp/core/ucp_ep.c

```diff
@@ -834,60 +842,87 @@ int ucp_ep_config_is_equal(const ucp_ep_config_key_t *key1,
-     * [ reg_cost.overhead + size * md_attr->reg_cost.growth + latency + overhead +
-     *   reg_cost.overhead + size * md_attr->reg_cost.growth + overhead + latency +
-     *   size/bw + latency + overhead + latency ]
-     *
-     * Isolating the 'size' yields the rndv_thresh.
-     * The used latency functions for eager_zcopy and rndv are also specified in
-     * the UCX wiki */
-
-    if (md_attr->cap.flags & UCT_MD_FLAG_REG) {
-        md_reg_growth   = md_attr->reg_cost.growth;
-        md_reg_overhead = md_attr->reg_cost.overhead;
-    } else {
-        md_reg_growth   = 0;
-        md_reg_overhead = 0;
+    ucp_context_h context = worker->context;
+    ucp_md_map_t md_map   = 0;
+    ucp_lane_index_t lane;
+    ucp_rsc_index_t rsc_index;
+    ucp_md_index_t md_index;
+    uct_md_attr_t *md_attr;
+    uct_iface_attr_t *iface_attr;
+    int i;
+
+    memset(params, 0, sizeof(*params));
+
+    /* latency and overhead used from lanes[0] (it is RTS/RTR/ATS values) */
+    rsc_index        = config->key.lanes[lanes[0]].rsc_index;
+    iface_attr       = ucp_worker_iface_get_attr(worker, rsc_index);
+    params->latency  = ucp_tl_iface_latency(context, iface_attr);
+    params->overhead = iface_attr->overhead;
+
+    for (i = 0; (lanes[i] != UCP_NULL_LANE) && (i < UCP_MAX_LANES); i++) {
+        lane      = lanes[i];
+        rsc_index = config->key.lanes[lane].rsc_index;
+        md_index  = config->md_index[lane];
+
+        if (!(md_map & UCS_BIT(md_index))) {
+            md_map |= UCS_BIT(md_index);
+            md_attr = &context->tl_mds[md_index].attr;
+            if (md_attr->cap.flags & UCT_MD_FLAG_REG) {
+                params->reg_growth   += md_attr->reg_cost.growth;
+                params->reg_overhead += md_attr->reg_cost.overhead;
+            }
+        }
+        params->bw += worker->ifaces[rsc_index].attr.bandwidth;
     }
+}
 
-    numerator = diff_percent * ((4 * ucp_tl_iface_latency(context, iface_attr)) +
-                (3 * iface_attr->overhead) +
-                (md_reg_overhead * (1 + recv_reg_cost))) -
-                 md_reg_overhead - iface_attr->overhead;
+static size_t ucp_ep_config_calc_rndv_thresh(ucp_worker_t *worker,
+                                             const ucp_ep_config_t *config,
+                                             const ucp_lane_index_t *eager_lanes,
+                                             const ucp_lane_index_t *rndv_lanes,
+                                             int recv_reg_cost)
+{
+    ucp_context_h context = worker->context;
+    double diff_percent   = 1.0 - context->config.ext.rndv_perf_diff / 100.0;
+    ucp_ep_thresh_params_t eager_zcopy;
+    ucp_ep_thresh_params_t rndv;
+    double numerator, denumerator;
+    ucp_rsc_index_t rsc_index;
+    uct_iface_attr_t *iface_attr;
+
+    /* All formulas and descriptions are listed at
+     * https://github.com/openucx/ucx/wiki/Rendezvous-Protocol-threshold-for-multilane-mode */
```

Comment:
Please link it to main FAQ page before merge

Reply:
this page is on Wiki. is it not ok?